File: src/typechecker/classes.rs
use super::*;

/// Builds `self.classes` in two passes over every `Stmt::ClassDef` in the
/// program. Pass one just registers each class's *name* with an empty,
/// placeholder `ClassInfo` — this exists purely so that pass two can
/// freely resolve a field/param/return type annotation that names *any*
/// class, including one declared later in the file or the class's own
/// name (self-referential fields, e.g. a `Node` with a `next: Node`
/// field). Pass two then does the real work for each class: resolves
/// every const/field annotation into a `Type` (consts and instance vars
/// are merged into one `fields` map, since both are read unqualified
/// inside methods via implicit `self` and the checker doesn't need to
/// tell them apart afterward), and resolves every method's param/return
/// types into a `FunctionSig` — but does not yet check any method
/// *bodies*; that happens later in `check_classes`, once every class's
/// signatures are known.
pub(super) fn collect_classes(
    checker: &mut TypeChecker,
    program: &[Stmt],
) -> Result<(), TypeError> {
    // Pre-register every class name (with an empty placeholder) first, so
    // field/param/return type annotations can name any class regardless
    // of declaration order (including a class referencing itself).
    for stmt in program {
        if let Stmt::ClassDef { name, .. } = stmt {
            checker.classes.entry(name.clone()).or_insert(ClassInfo {
                fields: HashMap::new(),
                methods: HashMap::new(),
            });
        }
    }

    for stmt in program {
        let Stmt::ClassDef {
            name,
            consts,
            fields,
            methods,
            ..
        } = stmt
        else {
            continue;
        };

        let mut field_types = HashMap::new();
        for c in consts {
            let Stmt::ConstDecl {
                name: cname,
                type_ann,
                line,
                column,
                ..
            } = c
            else {
                unreachable!("parser only ever puts ConstDecl in ClassDef.consts")
            };
            let Some(ann) = type_ann else {
                return Err(TypeError {
                    message: "class constants require an explicit type annotation".to_string(),
                    line: *line,
                    column: *column,
                });
            };
            let ty = checker.resolve_type(&ann.name, ann.line, ann.column)?;
            field_types.insert(cname.clone(), ty);
        }
        for f in fields {
            let ty = checker.resolve_type(&f.type_ann.name, f.line, f.column)?;
            field_types.insert(f.name.clone(), ty);
        }

        let mut method_sigs = HashMap::new();
        for m in methods {
            let Stmt::FunctionDef {
                name: mname,
                params,
                return_type,
                line,
                column,
                ..
            } = m
            else {
                unreachable!("parser only ever puts FunctionDef in ClassDef.methods")
            };
            let mut param_types = Vec::new();
            for p in params {
                param_types.push(checker.resolve_type(&p.type_ann.name, p.line, p.column)?);
            }
            let ret = match return_type {
                Some(t) => Some(checker.resolve_type(&t.name, *line, *column)?),
                None => None,
            };
            method_sigs.insert(
                mname.clone(),
                FunctionSig {
                    param_types,
                    return_type: ret,
                },
            );
        }

        checker.classes.insert(
            name.clone(),
            ClassInfo {
                fields: field_types,
                methods: method_sigs,
            },
        );
    }
    Ok(())
}

/// Type-checks every class method body (including `initializer`), with
/// the class's fields/consts pre-declared in the method's scope so bare
/// names resolve to them (implicit `self`, Ruby-ivar style).
pub(super) fn check_classes(checker: &mut TypeChecker, program: &[Stmt]) -> Result<(), TypeError> {
    for stmt in program {
        let Stmt::ClassDef {
            name,
            fields,
            methods,
            ..
        } = stmt
        else {
            continue;
        };
        check_fields_assigned_in_initializer(checker, name, fields, methods)?;
        let field_types = checker.classes[name].fields.clone();
        for m in methods {
            let Stmt::FunctionDef {
                params,
                body,
                return_type,
                ..
            } = m
            else {
                unreachable!("parser only ever puts FunctionDef in ClassDef.methods")
            };
            checker.push_scope();
            for (fname, fty) in &field_types {
                checker.declare_var(fname, fty.clone());
            }
            for p in params {
                let ty = checker.resolve_type(&p.type_ann.name, p.line, p.column)?;
                checker.declare_var(&p.name, ty);
            }
            let declared_return = match return_type {
                Some(t) => Some(checker.resolve_type(&t.name, t.line, t.column)?),
                None => None,
            };
            let actual_return = super::stmts::check_body_return_type(checker, body)?;
            if let (Some(declared), Some(actual)) = (&declared_return, &actual_return) {
                if declared != actual {
                    checker.pop_scope();
                    return Err(TypeError {
                        message: format!(
                            "method `{name}#{}` declared to return `{declared}`, but returns `{actual}`",
                            method_name(m)
                        ),
                        line: m.line(),
                        column: m.column(),
                    });
                }
            }
            checker.pop_scope();
        }
    }
    Ok(())
}

/// Requires every instance-variable declaration (a `FieldDecl` — always
/// valueless by construction) to be assigned somewhere in the class's
/// `initializer`, closing the soundness gap where reading a never-assigned
/// field type-checked as its declared type but evaluated to `Nil`.
///
/// The check is deliberately flow-insensitive: an assignment anywhere in
/// the initializer body (including inside an `if` branch or loop) counts,
/// so a conditionally-assigned field still passes. That over-approximation
/// is accepted — the goal is catching the common "declared but never set
/// anywhere" mistake, not full definite-assignment analysis. Inside
/// methods a bare `count = number` assignment parses as `Stmt::VarDecl`
/// (implicit self), so collecting `VarDecl` names is what detects field
/// assignment.
fn check_fields_assigned_in_initializer(
    _checker: &TypeChecker,
    class_name: &str,
    fields: &[crate::ast::FieldDecl],
    methods: &[Stmt],
) -> Result<(), TypeError> {
    if fields.is_empty() {
        return Ok(());
    }
    let initializer = methods.iter().find_map(|m| match m {
        Stmt::FunctionDef { name, body, .. } if name == "initializer" => Some(body),
        _ => None,
    });
    let mut assigned = std::collections::HashSet::new();
    if let Some(body) = initializer {
        collect_assigned_names(body, &mut assigned);
    }
    for f in fields {
        if !assigned.contains(f.name.as_str()) {
            return Err(TypeError {
                message: format!(
                    "field `{}` of class `{class_name}` is never assigned in `initializer` \
                     (it would be `Nil` at runtime, not `{}`)",
                    f.name, f.type_ann.name
                ),
                line: f.line,
                column: f.column,
            });
        }
    }
    Ok(())
}

/// Shared by `Expr::FieldAccess` and `Stmt::FieldAssign`: resolves
/// `object.field`'s type, erroring if `object` isn't a class instance or
/// the class has no such field.
pub(super) fn check_field_access(
    checker: &mut TypeChecker,
    object: &Expr,
    field: &str,
    line: usize,
    column: usize,
) -> Result<Type, TypeError> {
    let object_ty = super::exprs::check_expr(checker, object)?;
    let Type::Instance(class_name) = &object_ty else {
        return Err(TypeError {
            message: format!("cannot access field `{field}` on `{object_ty}`"),
            line,
            column,
        });
    };
    checker.classes[class_name]
        .fields
        .get(field)
        .cloned()
        .ok_or_else(|| TypeError {
            message: format!("class `{class_name}` has no field `{field}`"),
            line,
            column,
        })
}

/// Handles both `ClassName.new(args)` (when `object` is a bare `Ident`
/// naming a class rather than a bound variable) and ordinary
/// `instance.method(args)` calls.
pub(super) fn check_method_call(
    checker: &mut TypeChecker,
    object: &Expr,
    method: &str,
    args: &[Expr],
    line: usize,
    column: usize,
) -> Result<Type, TypeError> {
    if let Expr::Ident { name, .. } = object {
        if checker.lookup_var(name).is_none() && checker.classes.contains_key(name) {
            return check_construction(checker, name, method, args, line, column);
        }
    }

    let object_ty = super::exprs::check_expr(checker, object)?;
    let Type::Instance(class_name) = &object_ty else {
        return Err(TypeError {
            message: format!("cannot call method `{method}` on `{object_ty}`"),
            line,
            column,
        });
    };
    let sig = checker.classes[class_name]
        .methods
        .get(method)
        .cloned()
        .ok_or_else(|| TypeError {
            message: format!("class `{class_name}` has no method `{method}`"),
            line,
            column,
        })?;
    super::calls::check_call_args(
        checker,
        &format!("{class_name}#{method}"),
        &sig,
        args,
        line,
        column,
    )?;
    Ok(sig.return_type.unwrap_or(Type::Nil))
}

/// Type-checks `ClassName.new(args)`, dispatched from `check_method_call`
/// when `method_call`'s object is a bare `Ident` naming a known class
/// (rather than a variable bound to an instance). Only `new` is a valid
/// "static method" name — anything else is an error. If the class
/// declared an `initializer` method, arguments are checked against its
/// signature via `check_call_args`; if it didn't, `.new` must be called
/// with zero arguments (a helpful error otherwise). Always yields
/// `Type::Instance(class_name)` on success, since construction always
/// produces an instance of the class being constructed.
fn check_construction(
    checker: &mut TypeChecker,
    class_name: &str,
    method: &str,
    args: &[Expr],
    line: usize,
    column: usize,
) -> Result<Type, TypeError> {
    if method != "new" {
        return Err(TypeError {
            message: format!("class `{class_name}` has no static method `{method}`"),
            line,
            column,
        });
    }
    match checker.classes[class_name]
        .methods
        .get("initializer")
        .cloned()
    {
        Some(sig) => {
            super::calls::check_call_args(
                checker,
                &format!("{class_name}.new"),
                &sig,
                args,
                line,
                column,
            )?;
        }
        None if !args.is_empty() => {
            return Err(TypeError {
                message: format!(
                    "class `{class_name}` has no initializer, so `.new` takes no arguments"
                ),
                line,
                column,
            });
        }
        None => {}
    }
    Ok(Type::Instance(class_name.to_string()))
}

/// Collects every name assigned anywhere in `body` into `assigned`,
/// recursing through `if`/`while`/`for` bodies (flow-insensitively — a
/// conditional assignment still counts; see
/// `check_fields_assigned_in_initializer`). Both a bare `name = value`
/// (`Stmt::VarDecl`, how implicit-self field assignment parses inside
/// methods) and an explicit `object.field = value` (`Stmt::FieldAssign`)
/// register the assigned name.
fn collect_assigned_names(body: &[Stmt], assigned: &mut std::collections::HashSet<String>) {
    for stmt in body {
        match stmt {
            Stmt::VarDecl { name, .. } => {
                assigned.insert(name.clone());
            }
            Stmt::FieldAssign { field, .. } => {
                assigned.insert(field.clone());
            }
            Stmt::If {
                then_body,
                elsif_branches,
                else_body,
                ..
            } => {
                collect_assigned_names(then_body, assigned);
                for (_, branch) in elsif_branches {
                    collect_assigned_names(branch, assigned);
                }
                if let Some(body) = else_body {
                    collect_assigned_names(body, assigned);
                }
            }
            Stmt::While { body, .. } | Stmt::For { body, .. } => {
                collect_assigned_names(body, assigned);
            }
            _ => {}
        }
    }
}

/// Extracts a method's name from its `Stmt::FunctionDef` node, for building
/// the `"ClassName#method_name"` label used in return-type-mismatch error
/// messages in `check_classes`. Returns `"?"` for any other statement kind,
/// which should never actually happen since `ClassDef.methods` only ever
/// contains `FunctionDef`s (see the `unreachable!` guards in `collect_classes`
/// and `check_classes`).
fn method_name(method: &Stmt) -> &str {
    match method {
        Stmt::FunctionDef { name, .. } => name,
        _ => "?",
    }
}

File: ruby/lib/yara/typechecker/classes.rb
module Yara
  class TypeChecker
    private

    # Builds the class table in three passes: register every name (so any
    # annotation can name any class, itself included), fill each class's own
    # members, then flatten inheritance.
    #
    # @param program [Array<AST::Node>]
    # @return [void]
    def collect_classes(program)
      classes = program.grep(AST::ClassDef)
      classes.each { |klass| @classes[klass.name] ||= ClassInfo.new({}, {}) }
      classes.each do |klass|
        fields = {}
        klass.consts.each do |const|
          annotation = const.type_ann ||
                       type_error("type/class-const-requires-annotation", [], const.line, const.column)
          fields[const.name] = resolve_type(annotation.name, annotation.line, annotation.column)
        end
        klass.fields.each do |field|
          fields[field.name] = resolve_type(field.type_ann.name, field.line, field.column)
        end
        methods = {}
        klass.methods.each { |method| methods[method.name] = signature_of(method) }
        @classes[klass.name] = ClassInfo.new(fields, methods)
      end
      flatten_inheritance(classes)
    end

    # Merges each parent's (already flattened) members into its children,
    # parents first; a child's own member wins over an inherited one. Unknown
    # parents and cycles are reported at the child's `class`.
    #
    # @param classes [Array<AST::ClassDef>]
    # @return [void]
    def flatten_inheritance(classes)
      definitions = {}
      parents = {}
      classes.each do |klass|
        definitions[klass.name] = klass
        parents[klass.name] = klass.parent if klass.parent
      end
      parents.each do |child, parent|
        next if @classes.key?(parent)

        type_error("type/unknown-parent-class", [child, parent], definitions[child].line, definitions[child].column)
      end

      order = []
      @classes.each_key { |name| visit_for_cycle(name, parents, definitions, [], order) }
      order.each do |name|
        parent = parents[name]
        next unless parent

        child = @classes[name]
        child.fields = @classes[parent].fields.merge(child.fields)
        child.methods = @classes[parent].methods.merge(child.methods)
      end
    end

    # Depth-first walk up the parent chain, appending each class to `order`
    # after its ancestors. A cycle is reported at the first of its classes
    # the walk reached, where Rust picks one by `HashMap` order.
    #
    # @param name [String]
    # @param parents [Hash{String => String}]
    # @param definitions [Hash{String => AST::ClassDef}]
    # @param path [Array<String>] classes on the current chain
    # @param order [Array<String>]
    # @return [void]
    def visit_for_cycle(name, parents, definitions, path, order)
      return if order.include?(name)

      parent = parents[name]
      if parent
        if path.include?(parent) || parent == name
          type_error("type/inheritance-cycle", [parent], definitions[parent].line, definitions[parent].column)
        end
        visit_for_cycle(parent, parents, definitions, path + [name], order)
      end
      order << name
    end

    # Checks every method body with the class's fields pre-declared, since
    # methods read them unqualified.
    #
    # @param program [Array<AST::Node>]
    # @return [void]
    def check_classes(program)
      program.grep(AST::ClassDef).each do |klass|
        check_fields_assigned(program, klass)
        fields = @classes[klass.name].fields
        klass.methods.each do |method|
          @environment.push_scope
          fields.each { |name, type| @environment.declare(name, type) }
          method.params.each do |param|
            @environment.declare(param.name, resolve_type(param.type_ann.name, param.line, param.column))
          end
          check_return_type(method, "type/method-return-type-mismatch", [klass.name, method.name])
          @environment.pop_scope
        end
      end
    end

    # Every instance variable, own or inherited, must be assigned somewhere
    # in `initializer`. Flow-insensitive: an assignment inside an `if` counts.
    #
    # @param program [Array<AST::Node>]
    # @param klass [AST::ClassDef]
    # @return [void]
    def check_fields_assigned(program, klass)
      fields = klass.fields.dup
      parent = klass.parent
      while parent
        definition = program.find { |statement| statement.is_a?(AST::ClassDef) && statement.name == parent }
        break unless definition

        fields.concat(definition.fields)
        parent = definition.parent
      end
      return if fields.empty?

      initializer = klass.methods.find { |method| method.name == "initializer" }
      assigned = initializer ? assigned_names(initializer.body) : []
      fields.each do |field|
        next if assigned.include?(field.name)

        type_error("type/field-never-assigned", [field.name, klass.name, field.type_ann.name], field.line, field.column)
      end
    end

    # Every name assigned anywhere in `body`, nested blocks included: a bare
    # `name = value` (how implicit-self assignment parses) or `object.name = value`.
    #
    # @param body [Array<AST::Node>]
    # @return [Array<String>]
    def assigned_names(body)
      body.flat_map do |statement|
        case statement
        when AST::VarDecl then [statement.name]
        when AST::FieldAssign then [statement.field]
        when AST::If
          assigned_names(statement.then_body) +
            statement.elsif_branches.flat_map { |_, branch| assigned_names(branch) } +
            assigned_names(statement.else_body || [])
        when AST::While, AST::For then assigned_names(statement.body)
        else []
        end
      end
    end

    # @param object [AST::Node]
    # @param field [String]
    # @param line [Integer]
    # @param column [Integer]
    # @return [Type] the field's type
    def check_field_access(object, field, line, column)
      type = check_expr(object)
      type_error("type/cannot-access-field", [field, name_of(type)], line, column) if type.kind != "Instance"

      @classes[type.inner].fields[field] ||
        type_error("type/class-has-no-field", [type.inner, field], line, column)
    end

    # `ClassName.new(args)` when `object` names a class and no variable, an
    # instance method call on an instance, or a primitive method otherwise.
    #
    # @param object [AST::Node]
    # @param method [String]
    # @param args [Array<AST::Node>]
    # @param line [Integer]
    # @param column [Integer]
    # @return [Type]
    def check_method_call(object, method, args, line, column)
      if object.is_a?(AST::Ident) && !@environment.bound?(object.name) && @classes.key?(object.name)
        return check_construction(object.name, method, args, line, column)
      end

      type = check_expr(object)
      if type.kind != "Instance"
        return check_primitive_method(type, method, args, line, column) if type.receiver_kind

        type_error("type/cannot-call-method", [method, name_of(type)], line, column)
      end
      signature = @classes[type.inner].methods[method] ||
                  type_error("type/class-has-no-method", [type.inner, method], line, column)
      check_arguments("#{type.inner}##{method}", signature, args, line, column)
      signature.return_type || Type::NIL
    end

    # @param class_name [String]
    # @param method [String] must be `new`, possibly localized
    # @param args [Array<AST::Node>]
    # @param line [Integer]
    # @param column [Integer]
    # @return [Type]
    def check_construction(class_name, method, args, line, column)
      if @vocabulary.canonical_method(method) != "new"
        type_error("type/class-has-no-static-method", [class_name, method], line, column)
      end
      initializer = @classes[class_name].methods["initializer"]
      if initializer
        check_arguments("#{class_name}.new", initializer, args, line, column)
      elsif !args.empty?
        type_error("type/no-initializer-takes-no-args", [class_name], line, column)
      end
      Type.instance(class_name)
    end
  end
end
